test/unit_tests/arkode/CXX_serial/ark_test_adjoint_erk.cpp

```diff
@@ -0,0 +1,542 @@
+                            SUNAdjointCheckpointScheme checkpoint_scheme,
+                            const sunrealtype tf, const sunrealtype tout,
+                            N_Vector sf)
+{
+  sunrealtype t = tf;
+  SUNAdjointStepper_Evolve(adj_stepper, tout, sf, &t);
+
+  printf("Adjoint Solution:\n");
+  N_VPrint(sf);
+
+  printf("\nSUNAdjointStepper Stats:\n");
+  SUNAdjointStepper_PrintAllStats(adj_stepper, stdout, SUN_OUTPUTFORMAT_TABLE);
+  printf("\n");
+
+  return 0;
+}
+
+static void print_help(int argc, char* argv[], int exit_code)
+{
+  if (exit_code) { fprintf(stderr, "%s: option not recognized\n", argv[0]); }
+  else { fprintf(stderr, "%s ", argv[0]); }
+  fprintf(stderr, "options:\n");
+  fprintf(stderr, "--tf <real>         the final simulation time\n");
+  fprintf(stderr, "--dt <real>         the timestep size\n");
+  fprintf(stderr, "--order <int>       the order of the RK method\n");
+  fprintf(stderr, "--check-freq <int>  how often to checkpoint (in steps)\n");
+  fprintf(stderr, "--no-stages         don't checkpoint stages\n");
+  fprintf(stderr,
+          "--dont-keep         don't keep checkpoints around after loading\n");
+  fprintf(stderr, "--help              print these options\n");
+  exit(exit_code);
+}
+
+static void parse_args(int argc, char* argv[], ProgramArgs* args)
+{
+  for (int argi = 1; argi < argc; ++argi)
+  {
+    const char* arg = argv[argi];
+    if (!strcmp(arg, "--tf")) { args->tf = atof(argv[++argi]); }
+    else if (!strcmp(arg, "--dt")) { args->dt = atof(argv[++argi]); }
+    else if (!strcmp(arg, "--order")) { args->order = atoi(argv[++argi]); }
+    else if (!strcmp(arg, "--check-freq"))
+    {
+      args->check_freq = atoi(argv[++argi]);
+    }
+    else if (!strcmp(arg, "--no-stages")) { args->save_stages = SUNFALSE; }
+    else if (!strcmp(arg, "--dont-keep")) { args->keep_checks = SUNFALSE; }
+    else if (!strcmp(arg, "--help")) { print_help(argc, argv, 0); }
+    else { print_help(argc, argv, 1); }
+  }
+}
+
+int main(int argc, char* argv[])
+{
+  SUNContext sunctx = NULL;
+  SUNContext_Create(SUN_COMM_NULL, &sunctx);
+
+  // Since this a unit test, we want to abort immediately on any internal error
+  SUNContext_PushErrHandler(sunctx, SUNAbortErrHandlerFn, NULL);
+
+  ProgramArgs args;
+  args.tf          = SUN_RCONST(1.0);
+  args.dt          = SUN_RCONST(1e-4);
+  args.order       = 4;
+  args.save_stages = SUNTRUE;
+  args.keep_checks = SUNTRUE;
+  args.check_freq  = 2;
+  parse_args(argc, argv, &args);
+
+  //
+  // Create the initial conditions vector
+  //
+
+  sunindextype neq = 2;
+  N_Vector u       = N_VNew_Serial(neq, sunctx);
+  N_VConst(SUN_RCONST(1.0), u);
+
+  //
+  // Create the ARKODE stepper that will be used for the forward evolution.
+  //
+
+  const sunrealtype dt = args.dt;
+  sunrealtype t0       = SUN_RCONST(0.0);
+  sunrealtype tf       = args.tf;
+  const int nsteps     = (int)ceil(((tf - t0) / dt + 1));
```

Comment:
Another case where this looks off by one. I'm concerned the correctness will be susceptible to subtle rounding difference here versus in the time updating logic in ARKODE